runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/util/text/encoding/Encoding.kt

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package aws.smithy.kotlin.runtime.util.text.encoding
+
+import aws.smithy.kotlin.runtime.InternalApi
+
+private val ALPHA = (('A'..'Z') + ('a'..'z')).toSet()
+private val DIGIT = ('0'..'9').toSet()
+private val UNRESERVED = ALPHA + DIGIT + setOf('-', '.', '_', '~')
+private val SUB_DELIMS = setOf('!', '$', '&', '\'', '(', ')', '*', ',', ';', '=')
+private val VALID_UCHAR = UNRESERVED + SUB_DELIMS
+private val VALID_PCHAR = VALID_UCHAR + setOf(':', '@')
+private val VALID_FCHAR = VALID_PCHAR + setOf('/', '?')
+private val VALID_QCHAR = VALID_FCHAR - setOf('&', '=')
+
+@InternalApi
+public interface Encoding {
+    @InternalApi
+    public companion object {
+        public val UserInfo: Encoding = PercentEncoding("user info", VALID_UCHAR)
+        public val Path: Encoding = PercentEncoding("path", VALID_PCHAR)
+        public val Query: Encoding = PercentEncoding("query string", VALID_QCHAR, mapOf(' ' to '+'))
+        public val Fragment: Encoding = PercentEncoding("fragment", VALID_FCHAR)
+
+        internal val None = object : Encoding {
+            override val name = "(no encoding)"
+            override fun decode(encoded: String) = encoded
+            override fun encode(decoded: String) = decoded
+        }
+    }
+
+    public val name: String
+
+    public fun decode(encoded: String): String
+    public fun encode(decoded: String): String
+
+    public fun encodableFromDecoded(decoded: String): Encodable = Encodable(decoded, encode(decoded), this)
+    public fun encodableFromEncoded(encoded: String): Encodable {
+        val decoded = decode(encoded)
+        val reencoded = encode(decoded) // TODO is this right?
```

Comment:
That's an option and it's why I left the `TODO` in the code. My opinion is that we'd want to normalize any encoded representation we receive in case the encoding was flawed or non-standard.

For instance, if someone gives us the encoded string `"%61%62%63"`, its decoded representation is `"abc"` but its encoded representation _should also_ be `"abc"` since those characters don't require encoding. Worse yet, if someone gives us the encoded string `"💥"`, we'd want to reencode that as `"%F0%9F%92%A5"`.

I think the alternatives here are:
1. Re-encode any given encoded string to automatically normalize it (currently what the PR shows)
1. Accept the given encoded string as-is and accept that this may cause problems later on (e.g., during signing, in service responses, etc.)
1. Inspect the encoded string to verify its correctness, throw an exception if malformed (e.g., `"💥"`). If the encoded string is non-standardly encoded (e.g., `"%61%62%63"`):
    1. Throw an exception **–or–**
    1. Silently ignore

Reply:
Ok, it makes sense to me to re-encode values which don't require encoding ("abc") or re-encoding emojis to ensure downstream usages will work correctly.

It just stood out to me from an API perspective because I would expect that `"%61%62%63" == encodableFromEncoded("%61%62%63").encoded`. Can we log a message in cases where the re-encoded value is not the same as the given encoded value?

Reply:
Yes, I'll add a warning and also make clear in the KDocs that this will _reencode_ encoded text for normalization and validation purposes.